src/modules/ekf2/EKF/zero_gyro_update.cpp

```diff
@@ -0,0 +1,82 @@
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file zero_gyro_update.cpp
+ * Control function for ekf zero gyro update
+ */
+
+#include "ekf.h"
+
+void Ekf::controlZeroGyroUpdate(const imuSample &imu_delayed)
+{
+	// Fuse zero gyro at a limited rate
+	const bool zero_gyro_update_data_ready = isTimedOut(_time_last_zero_gyro_fuse, (uint64_t)2e5);
+
+	if (zero_gyro_update_data_ready) {
+		const bool continuing_conditions_passing = _control_status.flags.vehicle_at_rest
+				&& _control_status_prev.flags.vehicle_at_rest;
+
+		if (continuing_conditions_passing) {
+			// TODO: downsample to avoid aliasing
+			Vector3f delta_ang_scaled = (imu_delayed.delta_ang / imu_delayed.delta_ang_dt) * _dt_ekf_avg;
+			Vector3f innovation = _state.delta_ang_bias - delta_ang_scaled;
+
+			const float d_ang_sig = _dt_ekf_avg * math::constrain(_params.gyro_noise, 0.0f, 1.0f);
+			const float obs_var = sq(d_ang_sig);
+			Vector3f innov_var{
+				P(10, 10) + obs_var,
+				P(11, 11) + obs_var,
+				P(12, 12) + obs_var};
+
+			for (int i = 0; i < 3; i++) {
+				fuseDeltaAngBias(innovation(i), innov_var(i), i);
+			}
+
+			_time_last_zero_gyro_fuse = _time_delayed_us;
+		}
+	}
+}
+
+void Ekf::fuseDeltaAngBias(const float innov, const float innov_var, const int obs_index)
+{
+	Vector24f K;  // Kalman gain vector for any single observation - sequential fusion is used.
+	const unsigned state_index = obs_index + 10;
+
+	// calculate kalman gain K = PHS, where S = 1/innovation variance
+	for (int row = 0; row < _k_num_states; row++) {
+		K(row) = P(row, state_index) / innov_var;
+	}
+
+	measurementUpdate(K, innov_var, innov);
```

Comment:
Just for my own understanding, why no innovation check? We'd be screwed anyway with that bad gyro?

Reply:
Yes, an innovation filter means more parameters and we anyway want to always force fuse the data here because we know that if we're at rest, the gyro value is the bias, no matter if the rest of the filter agrees with that or not. We could argue that the innovation filter will be useful to reject outliers, but I don't think it'll be a real problem here.